### PR TITLE
[SPARK-22087][SPARK-14650][WIP][BUILD][REPL][CORE] Compile Spark REPL for Scala 2.12 + other 2.12 fixes

### PR DESCRIPTION
**Pull request by srowen**
## What changes were proposed in this pull request?

Enable Scala 2.12 REPL. Fix most remaining issues with 2.12 compilation and warnings, including:

- Selecting Kafka 0.10.1+ for Scala 2.12 and patching over a minor API difference
- Fixing lots of "eta expansion of zero arg method deprecated" warnings
- Resolving the SparkContext.sequenceFile implicits compile problem
- Fixing an odd but valid jetty-server missing dependency in hive-thriftserver

## How was this patch tested?

Existing tests

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 1]
```diff
@@ -2826,33 +2826,33 @@ object WritableConverter {
   // them automatically. However, we still keep the old functions in SparkContext for backward
   // compatibility and forward to the following functions directly.
 
-  implicit def intWritableConverter(): WritableConverter[Int] =
-    simpleWritableConverter[Int, IntWritable](_.get)
+  implicit val intWritableConverter: () => WritableConverter[Int] =
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: These changes were necessary to make the implicits work in 2.12 now that eta-expansion of zero-arg methods is deprecated and apparently doesn't work for implicit resolution. It passes MiMa, but we could be conservative and retain the existing methods, and add function vals instead.

File: core/src/main/scala/org/apache/spark/SparkContext.scala [hunk 2]
```diff
@@ -2826,33 +2826,33 @@ object WritableConverter {
...
   }
 
-  implicit def stringWritableConverter(): WritableConverter[String] =
-    simpleWritableConverter[String, Text](_.toString)
+  implicit val stringWritableConverter: () => WritableConverter[String] =
+    () => simpleWritableConverter[String, Text](_.toString)
 
-  implicit def writableWritableConverter[T <: Writable](): WritableConverter[T] =
-    new WritableConverter[T](_.runtimeClass.asInstanceOf[Class[T]], _.asInstanceOf[T])
+  implicit def writableWritableConverter[T <: Writable : ClassTag]: () => WritableConverter[T] =
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/SparkContext.scala]: ClassTag was required here, for reasons I don't fully get

File: core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala [hunk 3]
```diff
@@ -218,11 +218,13 @@ private[history] class FsHistoryProvider(conf: SparkConf, clock: Clock)
     if (!conf.contains("spark.testing")) {
       // A task that periodically checks for event log updates on disk.
       logDebug(s"Scheduling update thread every $UPDATE_INTERVAL_S seconds")
-      pool.scheduleWithFixedDelay(getRunner(checkForLogs), 0, UPDATE_INTERVAL_S, TimeUnit.SECONDS)
+      pool.scheduleWithFixedDelay(
+        getRunner(() => checkForLogs()), 0, UPDATE_INTERVAL_S, TimeUnit.SECONDS)
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/history/FsHistoryProvider.scala]: These changes avoid warnings about eta-expansion of zero-arg methods. It works fine in 2.11 as well; just not relying on syntactic sugar for the same.

File: core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala [hunk 4]
```diff
@@ -450,10 +450,9 @@ private[deploy] class Worker(
         }
       }(cleanupThreadExecutor)
 
-      cleanupFuture.onFailure {
-        case e: Throwable =>
-          logError("App dir cleanup failed: " + e.getMessage, e)
-      }(cleanupThreadExecutor)
+      cleanupFuture.failed.foreach(e =>
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/deploy/worker/Worker.scala]: onFailure, onSuccess are deprecated in 2.12. This should be equivalent in 2.11+2.12

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 5]
```diff
@@ -396,12 +396,12 @@ class DAGScheduler(
 
   /** Find ancestor shuffle dependencies that are not registered in shuffleToMapStage yet */
   private def getMissingAncestorShuffleDependencies(
-      rdd: RDD[_]): Stack[ShuffleDependency[_, _, _]] = {
-    val ancestors = new Stack[ShuffleDependency[_, _, _]]
+      rdd: RDD[_]): ArrayStack[ShuffleDependency[_, _, _]] = {
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: Stack is deprecated in 2.12 for poor performance; ArrayStack should work the same and be faster, in 2.11 too

File: core/src/main/scala/org/apache/spark/rdd/DoubleRDDFunctions.scala [hunk 6]
```diff
@@ -128,9 +128,9 @@ class DoubleRDDFunctions(self: RDD[Double]) extends Logging with Serializable {
     }
     // Compute the minimum and the maximum
     val (max: Double, min: Double) = self.mapPartitions { items =>
-      Iterator(items.foldRight(Double.NegativeInfinity,
-        Double.PositiveInfinity)((e: Double, x: (Double, Double)) =>
-        (x._1.max(e), x._2.min(e))))
+      Iterator(
+        items.foldRight((Double.NegativeInfinity, Double.PositiveInfinity)
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/rdd/DoubleRDDFunctions.scala]: More warnings about auto converting 2 args to a single tuple arg

File: core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala [hunk 7]
```diff
@@ -501,8 +501,8 @@ private class JavaIterableWrapperSerializer
 private object JavaIterableWrapperSerializer extends Logging {
   // The class returned by JavaConverters.asJava
   // (scala.collection.convert.Wrappers$IterableWrapper).
-  val wrapperClass =
-    scala.collection.convert.WrapAsJava.asJavaIterable(Seq(1)).getClass
+  import scala.collection.JavaConverters._
+  val wrapperClass = Seq(1).asJava.getClass
```
Comment (srowen) [inline core/src/main/scala/org/apache/spark/serializer/KryoSerializer.scala]: WrapAsJava is deprecated and this ought to be the equivalent to obtain a Java Iterable wrapper

File: core/src/test/scala/org/apache/spark/deploy/SparkSubmitUtilsSuite.scala [hunk 8]
```diff
@@ -83,11 +83,12 @@ class SparkSubmitUtilsSuite extends SparkFunSuite with BeforeAndAfterAll {
     val resolver = settings.getDefaultResolver.asInstanceOf[ChainResolver]
     assert(resolver.getResolvers.size() === 4)
     val expected = repos.split(",").map(r => s"$r/")
-    resolver.getResolvers.toArray.zipWithIndex.foreach { case (resolver: AbstractResolver, i) =>
-      if (1 < i && i < 3) {
-        assert(resolver.getName === s"repo-$i")
-        assert(resolver.asInstanceOf[IBiblioResolver].getRoot === expected(i - 1))
-      }
+    resolver.getResolvers.toArray.map(_.asInstanceOf[AbstractResolver]).zipWithIndex.foreach {
```
Comment (srowen) [inline core/src/test/scala/org/apache/spark/deploy/SparkSubmitUtilsSuite.scala]: This ends up being interpreted as a partial function, which wasn't the intent, and generates a warning. Should be equivalent

File: core/src/test/scala/org/apache/spark/scheduler/SchedulerIntegrationSuite.scala [hunk 9]
```diff
@@ -625,6 +625,8 @@ class BasicSchedulerIntegrationSuite extends SchedulerIntegrationSuite[SingleCor
           backend.taskFailed(taskDescription, fetchFailed)
         case (1, _, partition) =>
           backend.taskSuccess(taskDescription, 42 + partition)
+        case unmatched =>
```
Comment (srowen) [inline core/src/test/scala/org/apache/spark/scheduler/SchedulerIntegrationSuite.scala]: Many new warnings about a match not being exhaustive; this is one case where it was clearly avoidable. I left other instances alone

File: external/kafka-0-10-sql/pom.xml [hunk 10]
```diff
@@ -102,8 +102,19 @@
...
   <build>
     <outputDirectory>target/scala-${scala.binary.version}/classes</outputDirectory>
     <testOutputDirectory>target/scala-${scala.binary.version}/test-classes</testOutputDirectory>
   </build>
+
+  <profiles>
+    <profile>
+      <id>scala-2.12</id>
+      <properties>
+        <kafka.version>0.10.1.1</kafka.version>
```
Comment (srowen) [inline external/kafka-0-10-sql/pom.xml]: Only 0.10.1+ supports Scala 2.12. By the time a 2.12 build is actually supported we may be on to 0.10.2. Not sure. This at least makes it work

File: repl/scala-2.12/src/main/scala/org/apache/spark/repl/SparkILoop.scala [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+
+  /** Available commands */
+  override def commands: List[LoopCommand] = standardCommands
+
+  /**
+   * We override `createInterpreter` because we need to initialize Spark *before* the REPL
+   * sees any files, so that the Spark context is visible in those files. This is a bit of a
+   * hack, but there isn't another hook available to us at this point.
+   */
+  override def createInterpreter(): Unit = {
```
Comment (srowen) [inline repl/scala-2.12/src/main/scala/org/apache/spark/repl/SparkILoop.scala]: This is the only meaningful difference from the 2.11 REPL, as it has to hook into a different place. All other REPL-related code isn't specific to 2.11/2.12 and was moved out into the common src directory in the repl module

File: sql/core/src/test/scala/org/apache/spark/sql/DataFrameStatSuite.scala [hunk 12]
```diff
@@ -230,19 +230,16 @@ class DataFrameStatSuite extends QueryTest with SharedSQLContext {
 
     val resNaN1 = dfNaN.stat.approxQuantile("input1", Array(q1, q2), epsilon)
     assert(resNaN1.count(_.isNaN) === 0)
-    assert(resNaN1.count(_ == null) === 0)
```
Comment (srowen) [inline sql/core/src/test/scala/org/apache/spark/sql/DataFrameStatSuite.scala]: resNaN1 is an Array[Double] so can never contain null. This is always true and generated a warning

File: sql/core/src/test/scala/org/apache/spark/sql/streaming/EnsureStatefulOpPartitioningSuite.scala [hunk 13]
```diff
@@ -20,6 +20,7 @@ package org.apache.spark.sql.streaming
 import java.util.UUID
 
 import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.DataFrame
```
Comment (srowen) [inline sql/core/src/test/scala/org/apache/spark/sql/streaming/EnsureStatefulOpPartitioningSuite.scala]: This whole file fix is already going into master separately as a hotfix

File: sql/hive-thriftserver/pom.xml [hunk 14]
```diff
@@ -63,6 +63,16 @@
       <groupId>${hive.group}</groupId>
       <artifactId>hive-beeline</artifactId>
     </dependency>
+    <dependency>
```
Comment (srowen) [inline sql/hive-thriftserver/pom.xml]: For some reason, hive-thriftserver wouldn't compile, unable to find jetty.server classes. It has a point, these should explicit, even if it somehow worked before.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by srowen** [comment]
Good to see this passes, as I expected, because it means we can also commit this chunk of progress towards 2.12 without any downside for 2.11. 

Right now this means that the 2.12 build does compile, but I know it doesn't quite pass all tests (some weird paranamer problem, to start with)

The only part I was slightly uncertain about was the change to sequenceFile implicits, but MiMa hasn't complained. 

Worth a look at the changes, but I think this is worth committing before too long.

**Review by HyukjinKwon — COMMENTED** [review]
Up to my little knowledge, it looks good to me in general. What I checked was, to double check the comments left here and the same instances.

**Comment by srowen** [comment]
Merged to master

**Pull request closed without merging**